output/s3/s3_output.go

```diff
@@ -0,0 +1,190 @@
+
+	Working with S3 permissions: good luck; be advised that in many configurations,
+	accounts may be able to write data which they then instantly lose permission to
+	manage or even read back; this will result in permission denied errors from this
+	system, and may result in dirty state left behind (because we have to permission
+	to clean it up once we've created this, and cannot discover this until it's
+	too late!).  PRs for improving this behavior deeply welcome; this author
+	is not an S3/AWS/IAM professional.
+
+	Try the docs here: https://docs.aws.amazon.com/AmazonS3/latest/dev/example-bucket-policies.html --
+	these describe attaching policies to a bucket that can grant permissions to a user account.
+*/
+type Output struct {
+	spec          def.Output
+	hasherFactory func() hash.Hash
+}
+
+func New(spec def.Output) output.Output {
+	if spec.Type != Type {
+		panic(errors.ProgrammerError.New("This output implementation supports definitions of type %q, not %q", Type, spec.Type))
+	}
+	return &Output{
+		spec:          spec,
+		hasherFactory: sha512.New384,
+	}
+}
+
+func (o Output) Apply(basePath string) <-chan output.Report {
+	// We actually shell out to the entire streaming part of the tar system.
+	// All the formatting and hashing is identical; this just shoves the
+	//  stream to a S3 bucket instead of a local filesystem.
+	done := make(chan output.Report)
+	go func() {
+		defer close(done)
+		try.Do(func() {
+			// parse URI
+			u, err := url.Parse(o.spec.URI)
+			if err != nil {
+				panic(output.ConfigError.New("failed to parse URI: %s", err))
+			}
+			bucketName := u.Host
+			storePath := u.Path
+			var splay bool
+			switch u.Scheme {
+			case "s3":
+				splay = false
+			case "s3+splay":
+				splay = true
+			default:
+				panic(output.ConfigError.New("unrecognized scheme: %q", u.Scheme))
+			}
+
+			// load keys from env
+			// TODO someday URIs should grow smart enough to control this in a more general fashion -- but for now, host ENV is actually pretty feasible and plays easily with others.
+			keys, err := s3gof3r.EnvKeys()
+			if err != nil {
+				panic(S3CredentialsMissingError.Wrap(err))
+			}
+
+			// initialize writer to s3!
+			// if the URI indicated splay behavior, first stream data to {$bucketName}:{dirname($storePath)}/.tmp.upload.{basename($storePath)}.{random()};
+			// this allows us to start uploading before the final hash is determined and relocate it later.
+			// for direct paths, upload into place, because aws already manages atomicity at that scale (and they don't have a rename or copy operation that's free, because uh...?  no time to implement it since 2006, apparently).
+			putPath := storePath
+			if splay {
+				putPath = path.Join(path.Dir(storePath), ".tmp.upload."+path.Base(storePath)+"."+guid.New())
+			}
+			s3writer := makeS3writer(bucketName, putPath, keys)
+
+			// walk, fwrite, hash
+			o.spec.Hash = tar2.Save(s3writer, basePath, o.hasherFactory)
```

Comment:
For s3, what limitations are in the path? Your test mentions `repeatr-test/test-9/rt-splay/heap/`, and I see dots being used in `splay` mode. What I'm ruminating at is that we don't have a way to ask repeatr to gzip and save some 65% space on my jquery build, etc. Potentially, in non-splay mode we could look for an extension, but I'd be unhappy.

This hits our previous discussion about our unfortunate blindness of conflated storage vs transport, and should not be solved in this PR. So just for future reference, how does s3 limit storage URLs?